vault/logical_system.go

```diff
@@ -849,6 +851,118 @@ func (b *SystemBackend) handleRootPluginReloadUpdate(ctx context.Context, req *l
 	return &resp, nil
 }
 
+func (b *SystemBackend) handlePluginCatalogPinUpdate(ctx context.Context, _ *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	pluginType, pluginName, resp := requirePluginTypeAndName(d)
+	if resp != nil {
+		return resp, nil
+	}
+
+	pluginVersion, builtin, err := getVersion(d)
+	if err != nil {
+		return logical.ErrorResponse(err.Error()), nil
+	}
+	if pluginVersion == "" {
+		return logical.ErrorResponse("missing plugin version"), nil
+	}
+	if builtin {
+		return logical.ErrorResponse("cannot pin a builtin plugin: %q", pluginVersion), nil
```

Comment:
I'd like to support pinning to builtin versions, but it requires a bit of extra thought and logic because of the fact we don't want to store builtin versions in storage (they disappear/change as Vault gets updated).